### ISSUE TITLE
RareWordNormalisation

### ISSUE THREAD

**Issue by dupsys** [fchollet/keras#5383]
I want to use seq2seq prediction where the input sequence is "cu 2morr, gf" and the output sequence is "see you tomorrow, 
 girl friend". In this case the input and output seq lengths are different as shown below:
 
import matplotlib.pyplot as plt
plt.ion()
abbreviation, meaning = list(zip(*[(key, value) for key, value in {
    'ao': 'adult only',
    'u': 'you',
    'gf': 'girl friend',
    'sul': 'see you later',
    'tmi': 'too much information',
    'amz': 'amazing',
    'beg': 'big evil grim',
    'bff4l': 'best friend forever for life',
    '2mor': 'tomorrow',
    '2night': 'tonight',            
    '2oo': 'too',
    '4eva': 'forever',
    'abt': 'about',
    '': '',
}.iteritems()]))

# Define the symbol alphabet, including the padding symbol '#'
letters = '#'  +''.join(sorted({c for s in abbreviation+meaning for c in s}))

# A map from letters to indexes
letter_index = {c:i for i,c in enumerate(letters)}

# Compute the size of our alphabet
n_letters = len(letters)

# Compute maximum string length
max_length = max(map(len, abbreviation+meaning))

n_letters, max_length, letters

# model construction
word = Input((max_length,))
embedded_word = Embedding(32, n_letters, mask_zero=False)(word)
encoding = LSTM(128)(embedded_word)
repeated_encoding = RepeatVector(max_length)(encoding)
# We concatenate the repeated encoding with the embedded input string (in the last dimension).
merged_encoding = merge([repeated_encoding, embedded_word], mode='concat')
# Our LSTM decoder will produce a _sequence_ of 128-dimensional vectors.
decoded = LSTM(128, return_sequences=True)(merged_encoding)
# output_word = TimeDistributed(Dense(n_letters, activation='softmax'))(decoded)
output_word = TimeDistributed(Dense(n_letters, activation='softmax'))(decoded)

I am interested in making two models, one for abbreviation and the other for meaning. Do i need to add attention LSTM and how will i merge the two together, so that my output could be "see you tomorrow, 
 girl friend".

**Comment by unrealwill** [comment]
Hello,
Are you sure you want to use deep learning to do that?
Do a string split on your abbreviation sentences, convert each sub-string using your dictionary which maps abbreviation to word. Join. And be done with it.

You probably don't have enough data to infer the abbreviation - meaning dictionary from data. So the only additional value deep learning could bring is just typo correction, and it will probably do a bad job with it.

**Comment by dupsys** [comment]
Thank you. Will try your suggestion and see how it perform.

However, i read a papers about attention model and i quit understand the
theoretical paper but want to test the performance of the model with LSTM
model in Keras.

Below is my code in keras which i want to introduce attention, so that the
sentence can be modeled cohesively in the correct distribution of English
Language. I am using noisy dataset (e.g. tweet).  Where is the better layer
to introduce attention and will the following model be able to model
long-dependencies sentence meaningfully?

print('Building training model...')
hidden = 128
layersize = 128

model = Sequential()

model.add(GRU(hidden, return_sequences = True,
input_shape=(maxSequenceLength, len(char_2_id))))

model.add(TimeDistributed(Dense(layersize)))
model.add(TimeDistributed(Activation('relu')))
model.add(TimeDistributed(Dense(len(char_2_id))))
model.add(TimeDistributed(Activation('softmax')))

%time model.compile(loss='categorical_crossentropy', optimizer =
RMSprop(lr=0.001, rho=0.9, epsilon=1e-08, decay=0.0), metrics=['accuracy'])

print(model.summary())

# Test a simple prediction on a batch for this model.
print("Sample input Batch size:"),
print(inputChars[0:32, :, :].shape)
print("Sample input Batch labels (nextChars):"),
print(nextChars[0:32, :, :].shape)
outputs = model.predict(inputChars[0:32, :, :])
print("Output Sequence size:"),
print(outputs.shape)

Your humble soft suggestion will be a great help and assist to me.

Kind Regards
Abbey

*-----------------------------------------------------------*
*Abiodun MODUPE*
B.Sc (Hon's), MIT, CCNA,CCNP, PhD (Computer Science:In View)
School of Computer Science and Applied Mathematics
University of the Witwatersrand, Braamfontein
Johannesburg.
*"When you read you skin the surface when you study you discover the
treasure"*

**Comment by unrealwill** [comment]
The better position to introduce attention, is between encoder and decoder in a sequence-to-sequence (Graves) architecture, so that the decoder can peek selectively using attention into values from the encoding space (sometimes as in your case peeking into the encoding input directly can be advantageous), so that it doesn't have to 'interiorize' the whole encoded sequence to transfer it through the encoder-decoder bottleneck.
